### PR TITLE
feat(cookie): getter method to retrieve cookies

### PR DESCRIPTION
**Pull request by awtsui**
Added getter methods to SDK client to retrieve all cookies. 

Replacing #57 

[ACT-388](https://linear.app/valora/issue/ACT-388/sdk-add-ability-for-fc-sdk-to-expose-and-ingest-provider-session)

### CODE REVIEW

File: src/fiat-connect-client.ts [hunk 1]
```diff
@@ -140,6 +145,20 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
...
 
+      const headerSetCookie = response.headers.get('set-cookie')
+
+      if (headerSetCookie) {
+        headerSetCookie.split(',').forEach(async (cookie) => {
+          await this.cookieJar.setCookie(
+            Cookie.parse(cookie) ?? '',
+            this.config.baseUrl,
+            {
+              ignoreError: true,
```
Comment (jophish) [inline src/fiat-connect-client.ts:156]: What does `ignoreError` do here?

File: src/fiat-connect-client.ts [hunk 2]
```diff
@@ -140,6 +145,20 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
         return handleError(data)
       }
 
+      const headerSetCookie = response.headers.get('set-cookie')
+
+      if (headerSetCookie) {
+        headerSetCookie.split(',').forEach(async (cookie) => {
+          await this.cookieJar.setCookie(
```
Comment (jophish) [inline src/fiat-connect-client.ts:152]: What's the behavior of the resulting cookie jar if we end up setting the same cookie twice? For example, if we login twice, will the second time we set a previously-set cookieoverwrite the previous value?

File: src/fiat-connect-client.ts [hunk 3]
```diff
@@ -498,6 +517,10 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
       return handleError(error)
     }
   }
+
+  async getCookies(): Promise<String> {
```
Comment (jophish) [inline src/fiat-connect-client.ts:521]: Is the output of [getCookieString](https://github.com/salesforce/tough-cookie#getcookiestring) as easily deserializable as the output of [serialize](https://github.com/salesforce/tough-cookie#serializecallbackerr-serializedobject)?

File: src/index-node.ts [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
-import fetchCookie from 'fetch-cookie'
-export * from './types'
 
 export class FiatConnectClient extends FiatConnectClientImpl {
   constructor(
     config: FiatConnectClientConfig,
     signingFunction: (message: string) => Promise<string>,
   ) {
-    super(config, signingFunction, fetchCookie(fetch))
+    super(config, signingFunction, fetch)
```
Comment (jophish) [inline src/index-node.ts:10]: Revert the changes in this file?

File: src/fiat-connect-client.ts [hunk 5]
```diff
@@ -511,7 +534,7 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
  * it attempts to cast it to a string, put it in a ResponseError object, and
  * wrap it in a Result.err.
  **/
-function handleError<T>(error: unknown): Result<T, ResponseError> {
+export function handleError<T>(error: unknown): Result<T, ResponseError> {
```
Comment (jophish) [inline src/fiat-connect-client.ts:537]: Any reason we need to export this now?

**Review by jophish — COMMENTED** [review]
Some comments!

File: src/fiat-connect-client.ts (hunk 1 above)
Comment (awtsui) [reply src/fiat-connect-client.ts:156]: The IgnoreError flag catches error during parsing and does nothing. I was messing with it during testing, but it shouldn't be there.

File: src/fiat-connect-client.ts (hunk 2 above)
Comment (awtsui) [reply src/fiat-connect-client.ts:152]: If the cookie keys are the same, the previous cookie's value is overridden.

File: src/fiat-connect-client.ts (hunk 3 above)
Comment (awtsui) [reply src/fiat-connect-client.ts:521]: getCookieString can be deserialized with Cookie.parse, while serialize goes with deserialize. 

getCookieString returns 'key=value'
serialize returns a custom struct

What do you recommend?

File: src/fiat-connect-client.ts (hunk 5 above)
Comment (awtsui) [reply src/fiat-connect-client.ts:537]: Nope, should be reverted.

File: src/fiat-connect-client.ts [hunk 6]
```diff
@@ -140,6 +145,17 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
         return handleError(data)
       }
 
+      const headerSetCookie = response.headers.get('set-cookie')
+
+      if (headerSetCookie) {
+        headerSetCookie.split(',').forEach(async (cookie) => {
```
[the remark this answers is not in the document]
Comment (awtsui) [reply src/fiat-connect-client.ts]: I believe there is a chance that happens. Added a Promise.all block to contain the async functions.

File: test/index.test.ts [hunk 7]
```diff
@@ -939,4 +939,19 @@ describe('FiatConnect SDK', () => {
       )
     })
   })
+  describe('Cookie Handling', () => {
```
Comment (satish-ravi) [inline test/index.test.ts:942]: To stay consistent with other tests, we should update login tests to check if cookie jar is set and getCookies test just checks the return value from cookie jar. Although I am in process of refactoring login related stuff into a more generic SiweClient on #59, so I can address this on my PR.

File: src/types.ts [hunk 8]
```diff
@@ -56,6 +56,7 @@ export interface FiatConnectApiClient {
   getTransferStatus(
     params: TransferStatusRequestParams,
   ): Promise<Result<TransferStatusResponse, ResponseError>>
+  getCookies(): Promise<String>
```
Comment (satish-ravi) [inline src/types.ts:59]: use `string` over `String`

**Review by satish-ravi — APPROVED** [review]

**Review by jophish — APPROVED** [review]
